# Big Data Assignment — Reading Files, Cleaning, and Data Management
### Course: Algorithm Analysis (Gpo 104)
### Student: Kira Dayanna Espinoza Ortiz - A00841144
### Date: August 24, 2026

In [1]:
import pandas as pd
import numpy as np
## 1. Loading the Dataset
mis_datos = pd.read_csv("../data/cars.csv")
## Question 1. Review the structure


In [2]:
mis_datos.shape

(428, 15)

As we can see, the sahpe of my data set is 428 rows and 15 columns.

In [3]:
mis_datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 428 entries, 0 to 427
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Make         428 non-null    object 
 1   Model        428 non-null    object 
 2   Type         428 non-null    object 
 3   Origin       428 non-null    object 
 4   DriveTrain   428 non-null    object 
 5   MSRP         428 non-null    int64  
 6   Invoice      428 non-null    int64  
 7   EngineSize   425 non-null    float64
 8   Cylinders    425 non-null    float64
 9   Horsepower   423 non-null    float64
 10  MPG_City     426 non-null    float64
 11  MPG_Highway  426 non-null    float64
 12  Weight       423 non-null    float64
 13  Wheelbase    426 non-null    float64
 14  Length       427 non-null    float64
dtypes: float64(8), int64(2), object(5)
memory usage: 50.3+ KB


From this quick info check we can observe that some rows have less registered values than 428, which means we have missing values. We can also see the type of data registered in each row, in this data set, we have text objects, integers (whole numbers), and float numbers (with decimals).

In [4]:
mis_datos.columns

Index(['Make', 'Model', 'Type', 'Origin', 'DriveTrain', 'MSRP', 'Invoice',
       'EngineSize', 'Cylinders', 'Horsepower', 'MPG_City', 'MPG_Highway',
       'Weight', 'Wheelbase', 'Length'],
      dtype='object')

We can see the names of the 15 columns

In [5]:
mis_datos.head()

,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
0,Acura,MDX,SUV,Asia,All,36945,33337,3.5,6.0,265.0,17.0,23.0,4451.0,106.0,189.0
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,21761,2.0,4.0,200.0,24.0,31.0,2778.0,101.0,172.0
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,24647,2.4,4.0,200.0,22.0,29.0,3230.0,105.0,183.0
3,Acura,TL 4dr,Sedan,Asia,Front,33195,30299,3.2,6.0,270.0,20.0,28.0,NaN,108.0,186.0
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,39014,3.5,6.0,225.0,18.0,24.0,3880.0,115.0,197.0


Here we can observe the first rows of data

In [6]:
mis_datos.tail()

,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
423,Volvo,C70 LPT convertible 2dr,Sedan,Europe,Front,40565,38203,2.4,5.0,197.0,21.0,28.0,3450.0,105.0,186.0
424,Volvo,C70 HPT convertible 2dr,Sedan,Europe,Front,42565,40083,2.3,5.0,242.0,20.0,26.0,3450.0,105.0,186.0
425,Volvo,S80 T6 4dr,Sedan,Europe,Front,45210,42573,2.9,6.0,268.0,19.0,26.0,3653.0,110.0,190.0
426,Volvo,V40,Wagon,Europe,Front,26135,24641,1.9,4.0,170.0,22.0,29.0,2822.0,101.0,180.0
427,Volvo,XC70,Wagon,Europe,All,35145,33112,2.5,5.0,208.0,20.0,27.0,3823.0,109.0,186.0


And here the last rows of data.

In [7]:
mis_datos.sample(5)

,Make,Model,Type,Origin,DriveTrain,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
358,Subaru,Outback H6 4dr,Sedan,Asia,All,29345,26660,3.0,6.0,212.0,19.0,26.0,3610.0,104.0,184.0
408,Volkswagen,Passat GLS 4dr,Sedan,Europe,Front,23955,21898,1.8,4.0,170.0,22.0,31.0,3241.0,106.0,185.0
10,Audi,A4 3.0 Quattro 4dr manual,Sedan,Europe,All,33430,30366,3.0,6.0,220.0,17.0,26.0,3583.0,104.0,179.0
245,Mazda,MX-5 Miata convertible 2dr,Sports,Asia,Rear,22388,20701,1.8,4.0,142.0,23.0,28.0,2387.0,89.0,156.0
288,Mitsubishi,Outlander LS,SUV,Asia,Front,18892,17569,2.4,4.0,160.0,21.0,27.0,3240.0,103.0,179.0


We can also get a random sample of 5 rows from the dataset. 

In [8]:
mis_datos.describe()

,MSRP,Invoice,EngineSize,Cylinders,Horsepower,MPG_City,MPG_Highway,Weight,Wheelbase,Length
count,428.000000,428.000000,425.000000,425.000000,423.000000,426.000000,426.000000,423.000000,426.000000,427.000000
mean,32774.855140,30014.700935,3.195294,5.811765,215.576832,20.065728,26.840376,3578.229314,108.166667,186.381733
std,19431.716674,17642.117750,1.111932,1.557802,71.978098,5.249574,5.749505,759.034549,8.327582,14.369110
min,10280.000000,9875.000000,1.300000,3.000000,73.000000,10.000000,12.000000,1850.000000,89.000000,143.000000
25%,20334.250000,18866.000000,2.300000,4.000000,165.000000,17.000000,24.000000,3103.000000,103.000000,178.000000
50%,27635.000000,25294.500000,3.000000,6.000000,210.000000,19.000000,26.000000,3476.000000,107.000000,187.000000
75%,39205.000000,35710.250000,3.900000,6.000000,252.500000,21.750000,29.000000,3978.500000,112.000000,194.000000
max,192465.000000,173560.000000,8.300000,12.000000,500.000000,60.000000,66.000000,7190.000000,144.000000,238.000000


We can get a quick statistical summary of the data.

## Question 2 Check for missing values and replace them with the mean.

In this part, we first check how many missing values we have in each column.

In [9]:
mis_datos.isna().sum()

Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     3
Cylinders      3
Horsepower     5
MPG_City       2
MPG_Highway    2
Weight         5
Wheelbase      2
Length         1
dtype: int64

As we can see, we have several missing values, in EngineSize, Cylinders, Horsepower, MPG_City, MPG_Highway, Weight, Wheelbase, Length.
We will fill the missing values with the mean of each column.

In [11]:
missing_columns = mis_datos.columns[mis_datos.isna().sum() > 0]

for column in missing_columns:
    column_mean = mis_datos[column].mean()
    mis_datos[column] = mis_datos[column].fillna(column_mean)

mis_datos.isna().sum()

Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

After re-checking, now we have no missing values.

## Question 3 Which manufacturer produces the most vehicles, and which produces the fewest?

Now we will count how many vehicles are in each category of the "Make" column, in order to understand how many cars produces each manufacturer.

In [32]:
vehicles_by_manufacturer = mis_datos["Make"].value_counts()
vehicles_by_manufacturer

Make
Toyota           28
Chevrolet        27
Mercedes-Benz    26
Ford             23
BMW              20
Audi             19
Nissan           17
Honda            17
Chrysler         15
Volkswagen       15
Mitsubishi       13
Dodge            13
Volvo            12
Hyundai          12
Jaguar           12
Subaru           11
Kia              11
Pontiac          11
Mazda            11
Lexus            11
Buick             9
Mercury           9
Lincoln           9
Cadillac          8
GMC               8
Saturn            8
Suzuki            8
Infiniti          8
Acura             7
Saab              7
Porsche           7
Land Rover        3
Oldsmobile        3
Jeep              3
Isuzu             2
MINI              2
Scion             2
Hummer            1
Name: count, dtype: int64

Now we want the code to directly tell us who produces the most and the least vehicle models.

In [33]:
most_vehiclesm = vehicles_by_manufacturer.idxmax()
most_vehicles_count = vehicles_by_manufacturer.max()
fewest_vehiclesm = vehicles_by_manufacturer.idxmin()
fewest_vehicles_count = vehicles_by_manufacturer.min()
print(f"The manufacturer with the most vehicles: {most_vehiclesm} ({most_vehicles_count} models)")
print(f"The manufacturer with the fewest vehicles: {fewest_vehiclesm} ({fewest_vehicles_count} models)")

The manufacturer with the most vehicles: Toyota (28 models)
The manufacturer with the fewest vehicles: Hummer (1 models)


### Question 4 — Cars by region with front-wheel drive, rear-wheel drive, and all-wheel drive (All)

We need to summarize the cars with certain features (given in DriveTrain column) in each region.

In [34]:
summary_by_region_drivetrain = mis_datos.groupby([mis_datos["Origin"], mis_datos["DriveTrain"]])
print(summary_by_region_drivetrain.size())

Origin  DriveTrain
Asia    All           34
        Front         99
        Rear          25
Europe  All           36
        Front         37
        Rear          50
USA     All           22
        Front         90
        Rear          35
dtype: int64


As we can see Asia has the most vehicles with Front-wheels, Europe the most with all-wheels and the most with rear-wheels. While the US, has a mojority of front-wheels vehicles, but is not the leader in any category.

In [35]:
summary_by_region_drivetrain2 = pd.crosstab(mis_datos["Origin"], mis_datos["DriveTrain"])
summary_by_region_drivetrain2

DriveTrain,All,Front,Rear
Origin,,,
Asia,34,99,25
Europe,36,37,50
USA,22,90,35


We can also put this in a table, better organized.

### Question 5. The highest suggested retail price, the lowest suggested retail price, the average city miles per gallon, and the average highway miles per gallon.

In [36]:
highest_sprice = mis_datos["MSRP"].max()
lowest_sprice = mis_datos["MSRP"].min()

highest_car = mis_datos.loc[mis_datos["MSRP"].idxmax()]
lowest_car = mis_datos.loc[mis_datos["MSRP"].idxmin()]

average_mpg_city = mis_datos["MPG_City"].mean()
average_mpg_highway = mis_datos["MPG_Highway"].mean()

print(f"Highest MSRP: ${highest_sprice:,.2f} - {highest_car['Make']} {highest_car['Model']}")
print(f"Lowest MSRP: ${lowest_sprice:,.2f} - {lowest_car['Make']} {lowest_car['Model']}")

print(f"Average MPG (city): {average_mpg_city:.2f}")
print(f"Average MPG (highway): {average_mpg_highway:.2f}")

Highest MSRP: $192,465.00 - Porsche 911 GT2 2dr
Lowest MSRP: $10,280.00 - Kia Rio 4dr manual
Average MPG (city): 20.07
Average MPG (highway): 26.84


As we can see, the highest suggested retail price of /$192,465.00,  corresponds to the Porsche 911 GT2 2dr. And the lowest to a Kia Rio 4dr manual of /$10,280.00.
Another insight is that cars on average perform better on the highway (26.84 mpg), than on the city (20.07).

### Question 6. Select only the following variables: Make, Model, Type, Origin, DriveTrain, MSRP, Weight, Wheelbase, and Length. Keep only the observations where DriveTrain is equal to Front, and store this information in a variable called mis_nuevos_datos.

In [37]:
mis_nuevos_datos = mis_datos[["Make", "Model","Type", "Origin", "DriveTrain", "MSRP", "Weight", "Wheelbase", "Length" ]].copy()
mis_nuevos_datos = mis_nuevos_datos[mis_nuevos_datos["DriveTrain"] == "Front"].copy()
mis_nuevos_datos.shape

(226, 9)

Now the new data set only has 226 rows and the 9 columns asked. 226 of the original 428 cars are front-wheel drive, confirming it's the most common drivetrain, consistent with the Question 4 crosstab.

### Question 7
Print the first observations of the mis_nuevos_datos variable.
Using the information stored in mis_nuevos_datos, create three new variables:
• Weight_kilo, which indicates the weight of the vehicle in kilograms.
• Wheelbase_cm, which shows the value of the Wheelbase variable in centimeters.
• Length_cm, which indicates the length of the vehicle in centimeters.
Store all this information in the mis_nuevos_datos variable.

In [38]:
mis_nuevos_datos.head()

,Make,Model,Type,Origin,DriveTrain,MSRP,Weight,Wheelbase,Length
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,2778.000000,101.0,172.0
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,3230.000000,105.0,183.0
3,Acura,TL 4dr,Sedan,Asia,Front,33195,3578.229314,108.0,186.0
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,3880.000000,115.0,197.0
5,Acura,3.5 RL w/Navigation 4dr,Sedan,Asia,Front,46100,3893.000000,115.0,197.0


Here we can see a view of the first rows of the new data set.

In [39]:
LB_TO_KG = 0.453592
IN_TO_CM = 2.54

mis_nuevos_datos = mis_nuevos_datos.assign(
    Weight_kilo=mis_nuevos_datos["Weight"] * LB_TO_KG,
    Wheelbase_cm=mis_nuevos_datos["Wheelbase"] * IN_TO_CM,
    Length_cm=mis_nuevos_datos["Length"] * IN_TO_CM
)

mis_nuevos_datos.head()

,Make,Model,Type,Origin,DriveTrain,MSRP,Weight,Wheelbase,Length,Weight_kilo,Wheelbase_cm,Length_cm
1,Acura,RSX Type S 2dr,Sedan,Asia,Front,23820,2778.000000,101.0,172.0,1260.078576,256.54,436.88
2,Acura,TSX 4dr,Sedan,Asia,Front,26990,3230.000000,105.0,183.0,1465.102160,266.70,464.82
3,Acura,TL 4dr,Sedan,Asia,Front,33195,3578.229314,108.0,186.0,1623.056191,274.32,472.44
4,Acura,3.5 RL 4dr,Sedan,Asia,Front,43755,3880.000000,115.0,197.0,1759.936960,292.10,500.38
5,Acura,3.5 RL w/Navigation 4dr,Sedan,Asia,Front,46100,3893.000000,115.0,197.0,1765.833656,292.10,500.38


I created three new columns, reflecting new data derived from the original data set, which is useful as the typical measurements units we use in México are KG and Cm, not Lb and IN. However, the orihinal data is still there for comparison.

### Question 8
Using the information stored in mis_nuevos_datos, create a new variable called opinion
based on the suggested retail price.
• If the price is greater than 60,000, the value of opinion must be "Expensive".
• If the price is between 30,000 and 60,000, the value of opinion must be "Normal".
• If the price is less than 30,000, the value of opinion must be "Inexpensive".
Store the new variable in mis_nuevos_datos.

In [40]:
def opinion_p(price):
    if price > 60000:
        return "Expensive"
    elif price >= 30000:
        return "Normal"
    else:
        return "Inexpensive"

mis_nuevos_datos["opinion"] = mis_nuevos_datos["MSRP"].apply(opinion_p)

mis_nuevos_datos[["Make", "Model", "MSRP", "opinion"]].sample(10)

,Make,Model,MSRP,opinion
277,Mercury,Mountaineer,29995,Inexpensive
72,Chevrolet,Impala 4dr,21900,Inexpensive
369,Suzuki,Forenza S 4dr,12269,Inexpensive
239,MINI,Cooper S,19999,Inexpensive
388,Toyota,Camry Solara SE V6 2dr,21965,Inexpensive
212,Kia,Amanti 4dr,26000,Inexpensive
367,Suzuki,Aeno S 4dr,12884,Inexpensive
90,Chrysler,PT Cruiser 4dr,17985,Inexpensive
320,Pontiac,Sunfire 1SA 2dr,15495,Inexpensive
402,Volkswagen,Golf GLS 4dr,18715,Inexpensive


The new opinion column, now allow us to classify the price of each car, in order to facilitate the analysis. Here you can see a random sample of selected columns of the new data set.

In [41]:
mis_nuevos_datos["opinion"].value_counts()

opinion
Inexpensive    175
Normal          48
Expensive        3
Name: count, dtype: int64

We can also get the insight that there are 175 inexpensive cars, 48 normal, and 3 expensive.

### Conclusion
In conclusion, through the practice with this data set I was able to reinforce the processes learned in class, understanding how to read the data, clean it, summarize it, group it, filter it, transform it and classify it. It is incredible how easy it is to manipulate the data in a dataset once you understand the logic. I am sure that this knowledge is very valuable when analysing big data, to be able to process it more easily and get better insights. The easiest part for me was exploring the data, as the code is very simple; and the most challenging part were the last few questions, as they asked for more complex things, but once I checked the examples and understood the reasoning behind, everything was more clear. Compared to R, the program we used the last course, I am very excited to use more python, as I know it is more flexible and stronger than R for data analysis. Nevertheless, even if I noticed that some functions change, the logic is very similar. I hope I can take the most advantage of learning through this course in order to master the use of python.